### DISCUSSION TITLE
How to handle loading subsequent data without suspending?

### DISCUSSION THREAD

**Discussion by cevr** [bvaughn/suspense#15] [Q&A] [answered]
Hi there! 
I wish I had known about this library sooner, I've spent a couple weeks combing over the replay io codebase to implement my own suspense cache.

In my implementation, I opted for adding another status called 'revalidating' that basically allowed the cache refetch without suspending, but also provide information that the current value was now stale (and render some sort of pending state).

How would you handle this use case?

**Comment by bvaughn** [comment]
Oh that's interesting. Does it differ in any other significant ways besides the "revalidating" status? If not, it might be worth switching to this one 😄 

> How would you handle this use case?

I don't think I have a need to differentiate between _pending_ and "revalidating" as you call it– at least from an external "status" standpoint. For the component(s) _suspending_ to fetch the data, their behavior is controlled by whether the update is a transition or not. So the only difference...I think...would be if you wanted to show some different UI treatment between the first load and a reload?

**Reply by cevr** [reply]
In my example, I have a menu cache (which in terms of this library is a single entry cache) and an action later mutates the menu in the backend, so I would like to invalidate the menuCache afterwards but not suspend the component.


```ts
export type CacheSuspenseResponse<TValue> = {
  status: 'success' | 'revalidating';
  value: TValue;
  invalidate: () => void;
};

// ....
read(...cacheKey: TParams): CacheSuspenseResponse<TValue> {
    const cache = this; // maybe use context to grab cache boundaries?
    const value = cache.suspend(...cacheKey);
    const rerender = useRerender();
    // if it gets here, thats because the value was resolved successfully
    const lastStatus = React.useRef<CacheStatus>(STATUS_RESOLVED);

    React.useEffect(() => {
      const unsub = cache.subscribeToStatus((status) => {
        if (status !== lastStatus.current) {
          lastStatus.current = status;
          rerender();
        }
      }, ...cacheKey);
      return unsub;
    });

    return {
      status: value.stale ? 'revalidating' : 'success',
      value: value.value,
      invalidate: () => cache.invalidate(...cacheKey)
	}
}
```

The implementation of invalidate is basically the same as getOrCreateRecord

```tsx
async invalidate(...args: TParams): Promise<CacheStatusRecord<TValue> | undefined> {
    const cacheKey = this.getCacheKey(...args);
    let record = this.recordMap.get(cacheKey);
    if (record && record.status !== STATUS_PENDING) {
      const wakeable = createWakeable(`${this.debugLabel} ${cacheKey}}`);
      record = {
        status: CACHE_REVALIDATING,
        value: wakeable,
        cached: record.value,
      } as CacheStatusRecord<TValue>;

      this.recordMap.set(cacheKey, record);
      this.notifySubscribers(...args);

      await this.fetchAndStoreValue(record, ...args).catch(handleError);
      return record;
    }
  }
```

🙈 I'm still learning from the code I've been reading so pardon if any of this seems wrong

**Reply by cevr** [reply]
Heres the suspend code

```ts
  suspend(...args: [...TParams]): { value: TValue; stale: boolean } {
    const record = this.getOrCreateRecord(...args);
    if (record.status === STATUS_RESOLVED) {
      return { value: record.value, stale: false };
    }

    if (record.status === CACHE_REVALIDATING) {
      return { value: record.cached, stale: true };
    }

	// how else to recover from error boundary other than manually placing a cache.remove(key) in the error boundary
    if (record.status === STATUS_REJECTED) {
      this.scheduleEviction(...args);
    }

    throw record.value;
  }
```

**Comment by bvaughn** [comment]
> I would like to invalidate the menuCache afterwards but not suspend the component.

By "not suspend the component" do you mean "not show a fallback" or do you literally mean "not re-render the component"?

I assume you mean the first thing, which you can do by scheduling an update as a transition (`useTransition` / `startTransition`).

**Reply by cevr** [reply]
Haha thank you! I'll try this. Would i need some sort of state update so that it knows that the cache was evicted?

**Reply by bvaughn** [reply]
You could do that, although that might not be sufficient if multiple parts of your app consume that data. This is kind of what I'm exploring with #7 (but it's not done yet)

**Reply by cevr** [reply]
I've tried to implement it myself using `startTransition` but it doesn't seem to pause the suspense when initiated outside the component tree. Im subscribing to the cache status updates to notify the component, but it just causes it to suspend. Would you know how I can mark the suspense as a transition from an imperative process outside of react?

**Reply by bvaughn** [reply]
> Would you know how I can mark the suspense as a transition from an imperative process outside of react?

Schedule the state update within startTransition callback.

**Reply by cevr** [reply]
https://stackblitz.com/edit/react-ts-342kk5?file=App.tsx

I don't really know what I'm doing wrong, but I can't get the suspense to actually be marked as pending. It always suspends, whether i do it outside the suspending component or not.

**Reply by cevr** [reply]
Is it that notifying subscribers is not synchronous? From what I can gather it should be

**Reply by bvaughn** [reply]
I don't think I'm ready to answer these questions yet. It's basically what I'm trying to work out over on #7.

**Comment by bvaughn** [comment]
I have a solid lead on this (see this comment: https://github.com/bvaughn/suspense/issues/7#issuecomment-1454987568)

Here's a walk through with a demo:
https://www.loom.com/share/55b645c9730a4958b8a2f91dfc19b72f

**Reply by bvaughn** [reply]
Hit a snag with a React bug (https://github.com/facebook/react/issues/26316) but after the fix– I feel really good about the direction I'm exploring in #10. Have some documentation writing to do tomorrow, and maybe a bit more testing, but I hope to publish this as v0.0.12 in the next day or so.

**Reply by cevr** [reply]
Very cool, it was very interesting reading the code. I also wanted to thank you for making this repo, I'be been learning a lot.

I landed on an API that's something like this in my implementation:

```
const [value, revalidating] = Cache.read(id)
```

The reasoning was that I wanted to be able to specify a difference between a pending state because the key changed and a stale value.
I couldn't implement it with startTransition (no way to track the pending state if its outside react) in the non experimental version of react, but do you think it's possible with the new hooks?
I know mutations covered this use case for tracking an invalidation within the component after the mutation, but what about mutations that happen somewhere else in the tree?

**Reply by bvaughn** [reply]
> I couldn't implement it with startTransition (no way to track the pending state if its outside react) in the non experimental version of react, but do you think it's possible with the new hooks?
> I know mutations covered this use case for tracking an invalidation within the component after the mutation, but what about mutations that happen somewhere else in the tree?

I'm not sure if I'm understanding your question correctly, but if I am– I think the [`useCacheStatus`](https://suspense.vercel.app/useCacheStatus) hook could help here. It updates React components when cache status changes– whether the change was due to "suspense" (a read during render) or some external API call.

**Reply by cevr** [reply]
Does useCacheStatus allow us to track whether the component suspended because the key changed or because the cache considered the value stale? I'm wondering if that distinction is important

In a UI I can imagine that the distinction can be used to show different types of pending UIs (or non at all maybe) but I'm also not sure if that's even necessary haha. What are your thoughts?

**Reply by bvaughn** [reply]
> Does useCacheStatus allow us to track whether the component suspended because the key changed or because the cache considered the value stale? I'm wondering if that distinction is important

`useCacheStatus` requires your cache key(s) as params– so it only tells you about the status of a single entry.

If you want to know that the "key" changed (whatever that means– presumably some new prop?) then you could track that with a ref.

If you wanted to combine the two, maybe something like this?
```js
import { useEffect, useRef } from "react";
import { useCacheStatus, STATUS_PENDING } from "suspense";

function Example({ cacheKey }) {
  const prevCacheKey = useRef(cacheKey);
  useEffect(() => {
    prevCacheKey.current = cacheKey;
  });

  let status = useCacheStatus(cache, cacheKey);

  const didCacheKeyChange = cacheKey !== prevCacheKey.current;
  if (didCacheKeyChange && status === STATUS_PENDING) {
    status = "pending-key-change";
  }

  // ...
}
```

**Reply by cevr** [reply]
https://user-images.githubusercontent.com/34170625/223188990-63523f56-6818-4a20-ad7d-fec864928007.mov

Here's an example of what I mean. When there's no cache, the fallback is shown. When the next button is clicked, its wrapped in a startTransition and it shows a pending state to show that the next page is loading, and when the invalidate current page button is clicked, there is a small loading indicator in the top right that shows it's being "revalidated"

**Reply by bvaughn** [reply]
Sorry. Not sure how that applies to my most recent comment/suggestion. Does the info I mentioned not work for this use case somehow? May need you to spell it out for me :smile:

**Reply by cevr** [reply]
No I think you're right! Thanks for the dialog. I'll definitely be using this.

**Reply by bvaughn** [reply]
If you end up using this package in your app(s) I'd love to see a demo (or just hear how it worked out for you) 😄

**Reply by cevr** [reply]
Of course! Working on a v2 at work and want to build on top of suspense, so I'll be working a lot with this in the coming months 😃

**Comment by bvaughn** [comment, marked as answer]
Okay I think this thread can be "answered" now.

The general purpose answer for React is: Use a transition ([`useTransition`](https://beta.reactjs.org/reference/react/useTransition) or [`startTransition`](https://beta.reactjs.org/reference/react/startTransition)).

The specific answer for this package is: Use the new mutation hook ([`useCacheMutation`](https://suspense.vercel.app/useCacheMutation)) as shown in the ["mutating cache values"](https://suspense.vercel.app/guide/mutating-cache-values) example page.
